localization/beluga_demo_amcl3_localization/src/beluga_amcl3_demo.cpp

```diff
@@ -0,0 +1,466 @@
+            return;
+        }
+
+        const auto update_start_time = std::chrono::high_resolution_clock::now();
+        const auto new_estimate = particle_filter_->update(
+            base_pose_in_odom,  //
+            beluga_ros::SparsePointCloud3<float>{
+                cloud,
+                lidar_pose_in_base
+            });
+        const auto update_stop_time = std::chrono::high_resolution_clock::now();
+        const auto update_duration = update_stop_time - update_start_time;
+
+        if (new_estimate.has_value()) {
+            const auto& [base_pose_in_map, _] = new_estimate.value();
+            last_known_odom_transform_in_map_ = base_pose_in_map * base_pose_in_odom.inverse();
+            last_known_estimate_ = new_estimate.value();
+
+            RCLCPP_INFO(
+                rclcpp::get_logger("rclcpp"), "Particle filter update iteration stats: %ld particles - %.3fms",
+                particle_filter_->particles().size(), 
+                std::chrono::duration<double, std::milli>(update_duration).count());
+        }
+
+        // Transforms are always published to keep them current.
+        if (kTfBroadcast) {
+            if (last_known_odom_transform_in_map_.has_value()) {
+            auto message = geometry_msgs::msg::TransformStamped{};
+            // Sending a transform that is valid into the future so that odom can be used.
+            const auto expiration_stamp = tf2_ros::fromMsg(cloud->header.stamp) +
+                                            tf2::durationFromSec(kTransformTolerance);
+            message.header.stamp = tf2_ros::toMsg(expiration_stamp);
+            message.header.frame_id = kGlobalFrameId;
+            message.child_frame_id = kOdomFrameId;
+            message.transform = tf2::toMsg(*last_known_odom_transform_in_map_);
+            tf_broadcaster_->sendTransform(message);
+            }
+        }
+
+        // New pose messages are only published on updates to the filter.
+        if (new_estimate.has_value()) {
+            auto message = geometry_msgs::msg::PoseWithCovarianceStamped{};
+            message.header.stamp = cloud->header.stamp;
+            message.header.frame_id = kGlobalFrameId;
+            const auto& [base_pose_in_map, base_pose_covariance] = new_estimate.value();
+            tf2::toMsg(base_pose_in_map, message.pose.pose);
+            //tf2::covarianceEigenToRowMajor(base_pose_covariance, message.pose.covariance);
+            pose_pub_->publish(message);
+        }
+    }  
+
+    // Map displayer
+    // Adapted from:
+    // https://github.com/fzi-forschungszentrum-informatik/vdb_mapping_ros
+    void display_map(const openvdb::FloatGrid::Ptr grid) const {
+        visualization_msgs::msg::Marker marker_msg;
+        std_msgs::msg::ColorRGBA point_color;
+        openvdb::CoordBBox bbox = grid->evalActiveVoxelBoundingBox();
+        double min_z, max_z;
```

Comment:
@pvela2017 nit: better define and assign on the same line, then it can be `const`

Reply:
Changed in 64e5615601d05c975eb50e7cbd60ce2acc80b3b4